### ISSUE TITLE
look forward to your code

### ISSUE THREAD

**Issue by ogugugugugua** [huangsicong/TimbreTron#1]
hi, huangsicong, I've read the paper of your impressive work TimbreTron, the result is unbelievable! As I am doing similar research in this area, I am looking forward to see your code coming out! Bravo!

**Comment by zhuhaozh** [comment]
Look forward the codes too.

**Comment by rahulbhalley** [comment]
I have done my own implementation but am waiting for the official implementation as it would give better results. 

I hope you have refactored your PyTorch implementation by now. 😃

**Comment by khiner** [comment]
Also looking forward to seeing your implementation!

**Comment by amy-cao** [comment]
The result is awesome! Looking forward to the implementation!

**Comment by kureta** [comment]
Are there still plans to submit an implementation?

**Comment by huangsicong** [comment]
Yes. Sorry for the delay! We've been running into some copyright limitations so we won't be able to upload the dataset unfortunately. Then there were some issues with pytube and some Youtube links we used were gone. We'll only be able to upload the Youtube links that still work. We are still working on those issues and we'll upload the code after those issues are fixed.

**Comment by kureta** [comment]
No problem. I am grateful that you are working on it. Can you publish the code without data, so we can play with it until you sort out those issues?

**Comment by rahulbhalley** [comment]
Yes, having only code for now will work fine. We just want to experiment with the network's ability. Please make it available.

**Comment by kroaton** [comment]
Any news regarding this? Cheers.

**Comment by qedpi** [comment]
I'm exploring musical style transfer as well, and hope to explore your code soon :)

**Comment by nicholasnadel** [comment]
Also looking forward to your code / any updates!

**Comment by mikaelsouza** [comment]
Hello!

Any updates?

**Comment by timlod** [comment]
I'd also be interested in the code! Perhaps you can provide it without the dataset until those issues are solved?

**Comment by hiromu** [comment]
Hello, when will you upload the code?

**Comment by joshhansen** [comment]
I also would love to see the code.

On the other hand, it's been over a year so we may need to implement this independently. Tell me if my understanding is correct:

1) Convert source and target sounds into the frequency domain using CQT.
2) Train a CycleGAN to translate the CQT source and the CQT target between the two domains. The discriminators operate on the CQT representation.
3) Train a WaveNet to reconstruct target audio from the CQT representation.
4) For a given source audio sample, apply the CQT, use CycleGAN to translate to the target CQT domain, then apply your WaveNet model to translate it back to audio.

**Comment by pjcasas29** [comment]
> I also would love to see the code.
> 
> On the other hand, it's been over a year so we may need to implement this independently. Tell me if my understanding is correct:
> 
> 1. Convert source and target sounds into the frequency domain using CQT.
> 2. Train a CycleGAN to translate the CQT source and the CQT target between the two domains. The discriminators operate on the CQT representation.
> 3. Train a WaveNet to reconstruct target audio from the CQT representation.
> 4. For a given source audio sample, apply the CQT, use CycleGAN to translate to the target CQT domain, then apply your WaveNet model to translate it back to audio.

Yep these are the steps. The only thing is that in training the cyclegan you have to make sure to remove the checkerboard artifacts by replacing the Upsampling in the Generators to this:

https://distill.pub/2016/deconv-checkerboard/

Also include a gradient penalty for the discriminators. 

I tried to do this and the biggest challenge was in training the Wavenet model so had to use something else. Let me know if you have any success with it!

**Comment by joshhansen** [comment]
> > I also would love to see the code.
> > On the other hand, it's been over a year so we may need to implement this independently. Tell me if my understanding is correct:
> > 
> > 1. Convert source and target sounds into the frequency domain using CQT.
> > 2. Train a CycleGAN to translate the CQT source and the CQT target between the two domains. The discriminators operate on the CQT representation.
> > 3. Train a WaveNet to reconstruct target audio from the CQT representation.
> > 4. For a given source audio sample, apply the CQT, use CycleGAN to translate to the target CQT domain, then apply your WaveNet model to translate it back to audio.
> 
> Yep these are the steps. The only thing is that in training the cyclegan you have to make sure to remove the checkerboard artifacts by replacing the Upsampling in the Generators to this:
> 
> https://distill.pub/2016/deconv-checkerboard/
> 
> Also include a gradient penalty for the discriminators.
> 
> I tried to do this and the biggest challenge was in training the Wavenet model so had to use something else. Let me know if you have any success with it!

Thanks for those pointers---I'll keep you updated if I give it a try!

BTW, if you were hung up on learning the CQT inverse, [apparently](https://github.com/librosa/librosa/issues/165) this [MatLab code](https://code.soundsoftware.ac.uk/projects/constant-q-toolbox) includes an (approximate?) inverse of CQT---I wonder if that might be more effective than having to learn the inverse from data.

**Comment by ARTORGTimFi** [comment]
i would be happy to see the code, as promised in the paper

**Comment by bosnyan** [comment]
I have been following the thread for quite a while now and I must say that I am slightly disappointed in not seeing any code being published even after over a year of public demand for it. For anyone interested in timbre transfer, Lu et. al have [developed a timbre transfer system](https://github.com/ChienYuLu/Play-As-You-Like-Timbre-Enhanced-Multi-modal-Music-Style-Transfer) as well. 
It is an adaptation NVIDIA's MUNIT system, which is used for image-to-image transformation. The audio domain has essentially been adjusted to fit the system by translating waveforms to spectrograms. I have played around with this system a few months ago when orientating for my master thesis and I can at least say that it does work. There is quite an absolute lack of documentation and several things (most notable the data input and output transformations) have some parts that are hard coded and need adjusting to your situation. If you do manage to get up and running however, it should work just fine. 
Keep in mind that this system performs an one-to-one transformation. This means that you have to train each instrument pair (e.g. guitar and piano) separately such to transform between them. Furthermore, the system is ran over a training set where the instrument pair is separated in a folder Train A and Train B. The folders Test A and Test B contain the audio files which you wish to use for timbre transformation. 
So let us say that instrument A is the piano and instrument B is the guitar. Audio samples of both instruments are contained in their respective training folders. After training, an "image transformation" can be performed on the spectrograms from A to B or vice versa such that the contents of A are in the style of B.

As far as samples are concerned, there is no limit in how many training examples you can throw at the system. The more data, the better in fact. A great source for instrument samples by the way is [Wikiloops](https://www.wikiloops.com/backingtrack.php) in my opinion. It is an online community where the users "jam" together by recording their own tracks and publishing these online. Many high quality, isolated recordings can be found on here. There is a subscription fee for non-artists however and it sits at €15 per 3 months.

**Comment by gladys0313** [comment]
Hello! Any plan to upload the code now-ish?

**Comment by quangntran** [comment]
> It is an adaptation NVIDIA's MUNIT system, which is used for image-to-image transformation. The audio domain has essentially been adjusted to fit the system by translating waveforms to spectrograms. I have played around with this system a few months ago when orientating for my master thesis and I can at least say that it does work. There is quite an absolute lack of documentation and several things (most notable the data input and output transformations) have some parts that are hard coded and need adjusting to your situation. If you do manage to get up and running however, it should work just fine.
> Keep in mind that this system performs an one-to-one transformation. This means that you have to train each instrument pair (e.g. guitar and piano) separately such to transform between them. Furthermore, the system is ran over a training set where the instrument pair is separated in a folder Train A and Train B. The folders Test A and Test B contain the audio files which you wish to use for timbre transformation.

Thanks for sharing, that is super helpful to me!

**Comment by rahulbhalley** [comment]
Yes @quangntran TimbreTron has two one-to-one generative (or transformation) networks.

**Comment by quangntran** [comment]
> I have done my own implementation but am waiting for the official implementation as it would give better results.
> 
> I hope you have refactored your PyTorch implementation by now. 😃
Hi @rahulbhalley , is it possible you shared with us your implementation please?

**Comment by RussellSB** [comment]
Hi everyone, if people would like something similar, feel free to check out this: https://github.com/RussellSB/tt-vae-gan

While TimbreTron derives from CycleGAN, our approach derives from UNIT (a similar approach, that also considers variational encoding)

**Comment by diggerdu** [comment]
Any updates in year 2019+3?

**Comment by Alexlly1** [comment]
hi, huangsicong, I've read the paper of your impressive work TimbreTron, the result is unbelievable! As I am doing similar research in this area, I am looking forward to see your code coming out!

**Comment by Alexlly1** [comment]
> I have done my own implementation but am waiting for the official implementation as it would give better results.
> 
> I hope you have refactored your PyTorch implementation by now. 😃

Did you succeed？I would be very happy if you could provide the code
